helix-core/src/main/java/org/apache/helix/cloud/topology/VirtualTopologyGroupStrategy.java

```diff
@@ -0,0 +1,80 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.helix.util.HelixUtil;
+
+public enum VirtualTopologyGroupStrategy {
+
+  /**
+   * A default assignment strategy that is deterministic and stable where:
+   * 1. assignment is guaranteed consistent for same inputs.
+   * 2. number of instance shuffles is reduced.
+   */
+  DEFAULT {
+    @Override
+    public Map<String, Set<String>> computeAssignment(
+        int numGroups, String virtualGroupName, Map<String, Set<String>> zoneMapping) {
+      List<String> sortedInstances = HelixUtil.sortAndFlattenZoneMapping(zoneMapping);
+      Map<String, Set<String>> assignment = new HashMap<>();
+      int instancesPerGroup = sortedInstances.size() / numGroups;
+
+      // instances.size = instancePerGroup * numGroups + residual
+      // step 1, continuously assign following groupInd order until entering residual section
+      for (int instanceInd = 0; instanceInd < instancesPerGroup * numGroups; instanceInd++) {
+        int groupIndex = instanceInd / instancesPerGroup;
+        String groupId = computeVirtualGroupId(groupIndex, virtualGroupName);
+        assignment.putIfAbsent(groupId, new HashSet<>());
+        assignment.get(groupId).add(sortedInstances.get(instanceInd));
+      }
+      // step 2, assign the residuals, either stick to the last assigned group or round-robin to all groups.
+      for (int instanceInd = instancesPerGroup * numGroups; instanceInd < sortedInstances.size(); instanceInd++) {
+        int groupIndex = numGroups <= zoneMapping.keySet().size()
+            ? numGroups - 1
+            : instanceInd % numGroups;
+        String groupId = computeVirtualGroupId(groupIndex, virtualGroupName);
+        assignment.putIfAbsent(groupId, new HashSet<>());
+        assignment.get(groupId).add(sortedInstances.get(instanceInd));
+      }
+      return ImmutableMap.copyOf(assignment);
+    }
+  };
+
+  /**
+   * Compute the assignment for each virtual topology group.
+   *
+   * @param numGroups number of the virtual groups
+   * @param virtualGroupName virtual group name
+   * @param zoneMapping current zone mapping from zoneId to instanceIds
+   * @return the assignment as mapping from virtual group ID to instanceIds
+   */
+  public abstract Map<String, Set<String>> computeAssignment(
+      int numGroups, String virtualGroupName, Map<String, Set<String>> zoneMapping);
+
+  private static String computeVirtualGroupId(int groupIndex, String virtualGroupName) {
+    return virtualGroupName + "_" + groupIndex;
```

Comment:
nit, avoid hard code?

Reply:
What's your suggestion for constructing the virtual groupId? I can go with a string template for naming, but don't think that's much difference than this.

Reply:
I can see you did 

> private static final String DOMAIN_FIELD_SPLITTER = ",";
> private static final String DOMAIN_VALUE_JOINER = "=";

for the InstanceConfig. Then why hardcode here? It's better to keep the same style.